Introduce layouts (formerly widgets) to prepare for redesign

mmilata opened the pull request:
Most changes in this PR come from #1312, except there are no UI bits for T1, and the layouts have been tweaked to pass most UI tests.

Changes:
* UI components moved to model-specific directories, for some of them the model-independent code has been factored out
* some dialogs have been converted to the `Layout` abstraction (formerly `Widget`)

To be done:
* convert all  `app.*` code to use `Layout`s, remove direct imports of `trezor.ui.components.tt.*`
* update `ButtonRequest` message to include string field "type" (currently it is just written to log)

There are some spacing differences in the [UI diff](https://satoshilabs.gitlab.io/-/trezor/trezor-firmware/-/jobs/956622686/artifacts/master_diff/index.html), if these are a problem they can probably be worked around.

matejcik commented on core/src/apps/tezos/get_public_key.py:
```diff
@@ -1,7 +1,7 @@
 from trezor.messages import ButtonRequestType
 from trezor.messages.TezosPublicKey import TezosPublicKey
-from trezor.ui.text import Text
+from trezor.ui.components.tt.text import Text
```
couldn't we use `show_pubkey` here too?

mmilata replied:
Not sure if I overlooked it or didn't use it because of slight difference ("..." at the end if pubkey doesn't fit screen). We definitely should unify those (and make them scrollable if the pubkey doesn't fit), fixed in 48190477e1de3bf7ceca76880c7dcc587819fe42.

matejcik commented on core/src/trezor/ui/components/tt/confirm.py:
```diff
@@ -38,8 +30,9 @@ def __init__(
     ) -> None:
-        super().__init__()
         self.content = content
+        button_confirm = None  # type: Optional[Button]
```
please use the new-style type annotations: `button_confirm: Optional[Button] = None`

mmilata replied:
Fixed in 81b9c568ba0a0f3e9e070a6d56990094d2937857. `# type: ignore` can't be expressed in the new style, right?

matejcik replied:
correct; it's not a type, it's an instruction to the typechecker

matejcik commented on core/src/trezor/ui/layouts/tt.py:
```diff
@@ -0,0 +1,410 @@
+
+
+def _show_xpub(xpub: str, desc: str, cancel: str) -> Paginated:
+    pages = []  # type: List[ui.Component]
```
please use the new-style type annotation

mmilata replied:
Fixed in 379440b0a6baa70f6bebf4542e3fdee21c6f092c.

matejcik reviewed (changes requested):
minor comments

overall: why did you choose to have `show_pubkey` accept str instead of bytes?
not saying it's wrong (i did the same in `confirm_hex` or something), just wondering about reasoning and whether in case of `show_pubkey` the hexlification could be done in the layout; we _know_ that pubkey is bytes

otherwise basically LGTM, let's wait until the release branch, and then rebase&merge

mmilata commented:
Addressed the comments and rebased to master. Conflicts:
* `apps/bitcoin/get_address.py` (xpub magic, "others" -> "cosigner")
* `apps/bitcoin/sign_tx/layout.py` (amount units)
* `apps/cardano/layout.py` (map account paths to account numbers) - rolled changes back to what's in master

mmilata commented:
Rebased on master to resolve conflicts in fixtures. Ready to merge? [UI diff](https://satoshilabs.gitlab.io/-/trezor/trezor-firmware/-/jobs/1015469580/artifacts/master_diff/index.html)

matejcik reviewed (approved):
approval from me

but pinging @tsusanka, this is big enough that i'd like another pair of eyes

tsusanka reviewed (approved):
LGTM! I have skimmed through the code and discussed few things on-call with @mmilata. Let's merge it! :tada:

tsusanka reviewed (approved):
LGTM! I have skimmed through the code and discussed few things with @mmilata on-call. Let's merge this!

tsusanka reviewed (approved):
LGTM! I have skimmed through the code and discussed few things with @mmilata on-call. Let's merge this!

tsusanka commented:
Due to Github's 500 I have approved 3x times :roll_eyes:.

mmilata commented:
Thanks a lot for reviewing this thing guys! @matejcik can you please look at this little fix: f9f91a2b71a82d0d369e2339c5f64c86c9660e18?  Without it the Cancel button on get_public_key confirm doesn't actually cancel the operation.

matejcik commented:
> Thanks a lot for reviewing this thing guys! @matejcik can you please look at this little fix: [f9f91a2](https://github.com/trezor/trezor-firmware/commit/f9f91a2b71a82d0d369e2339c5f64c86c9660e18)? Without it the Cancel button on get_public_key confirm doesn't actually cancel the operation.

sure, LGTM.
(i mean it is not meaningfully changing anything if the Cancel does not work here but, well)

mmilata commented:
Rebased and fixed a typing issue exposed by bf562cfd4bc48f5e6ab4f3f7a4b4f3e63b6dfb59.

> (i mean it is not meaningfully changing anything if the Cancel does not work here but, well)

The Cancel *works* in current master but was broken in this branch which was fixed. Or am I misunderstanding you?

matejcik commented:
it works, but it's meaningless on a GetPublicKey message, because a potential attacker can obtain the pubkey silently anyway. I suppose we should remove the cancel button in the redesign.
